pkg/util/provider/machinecontroller/node.go

```diff
@@ -6,49 +6,280 @@
+		return
+	}
+
+	if node.DeletionTimestamp != nil {
+		delRetry, err := c.triggerMachineDeletion(context.Background(), node.Name)
+		if err != nil {
+			c.enqueueNodeAfter(node, time.Duration(delRetry), fmt.Sprintf("handling node UPDATE event. node %q marked for deletion", node.Name))
+		}
 		return
 	}
-	c.nodeQueue.Add(key)
+
+	c.enqueueNode(node, "handling ADD event for node")
 }
 
-func (c *controller) nodeUpdate(newObj any) {
-	c.nodeAdd(newObj)
+func (c *controller) updateNode(oldObj, newObj any) {
+	oldNode := oldObj.(*corev1.Node)
+	node := newObj.(*corev1.Node)
+	if oldNode == nil || node == nil {
+		klog.Errorf("Couldn't convert to node from object")
+		return
+	}
+
+	// If NotManagedByMCM annotation is present on node, don't process this node object
+	if _, annotationPresent := node.ObjectMeta.Annotations[machineutils.NotManagedByMCM]; annotationPresent {
+		return
+	}
+
+	machine, err := c.getMachineFromNode(node.Name)
+	if err != nil {
+		klog.Errorf("Unable to handle update event for node %s, couldn't fetch associated machine. Error: %s", node.Name, err)
+		return
+	}
+
+	// delete the machine if the node is deleted
+	if node.DeletionTimestamp != nil {
+		delRetry, err := c.triggerMachineDeletion(context.Background(), node.Name)
+		if err != nil {
+			c.enqueueNodeAfter(node, time.Duration(delRetry), fmt.Sprintf("handling node UPDATE event. backing node %q marked for deletion", node.Name))
+		}
+		return
+	}
+
+	// to reconcile on addition/removal of essential taints in machine lifecycle, example - critical component taint
+	if addedOrRemovedEssentialTaints(oldNode, node, machineutils.EssentialTaints) {
+		c.enqueueMachine(machine, fmt.Sprintf("handling node UPDATE event. atleast one of essential taints on backing node %q has changed", getNodeName(machine)))
+		return
+	}
+	if inPlaceUpdateLabelsChanged(oldNode, node) {
+		c.enqueueMachine(machine, fmt.Sprintf("handling node UPDATE event. in-place update label added or updated for node %q", getNodeName(machine)))
+		return
+	}
+	isMachineCrashLooping := machine.Status.CurrentStatus.Phase == v1alpha1.MachineCrashLoopBackOff
+	isMachineTerminating := machine.Status.CurrentStatus.Phase == v1alpha1.MachineTerminating
+	_, _, nodeConditionsHaveChanged := nodeConditionsHaveChanged(machine.Status.Conditions, node.Status.Conditions)
+
+	// Enqueue machine if node conditions have changed and machine is not in crashloop or terminating state
+	if nodeConditionsHaveChanged && !(isMachineCrashLooping || isMachineTerminating) {
+		c.enqueueMachine(machine, fmt.Sprintf("handling node UPDATE event. conditions of backing node %q differ from machine status", node.Name))
+	}
 }
 
-func (c *controller) nodeDelete(obj any) {
-	node, _ := obj.(*v1.Node)
-	if node == nil {
+func (c *controller) deleteNode(obj any) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Couldn't get object from tombstone %+v", obj)
```

Comment:
Please change to lower case
```suggestion
			klog.Errorf("couldn't get object from tombstone %+v", obj)
```